common/hashing_test.cpp

```diff
@@ -0,0 +1,589 @@
+                sizeof(T) == 8);
+  os << llvm::formatv(sizeof(T) == 1   ? "{0:x2}"
+                      : sizeof(T) == 2 ? "{0:x4}"
+                      : sizeof(T) == 4 ? "{0:x8}"
+                                       : "{0:x16}",
+                      static_cast<uint64_t>(value));
+}
+
+template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+auto operator<<(llvm::raw_ostream& os, HashedValue<T> hv)
+    -> llvm::raw_ostream& {
+  os << "hash " << hv.hash << " for value ";
+  PrintFullWidthHex(os, hv.v);
+  return os;
+}
+
+template <typename T, typename U,
+          typename = std::enable_if_t<std::is_integral_v<T>>,
+          typename = std::enable_if_t<std::is_integral_v<U>>>
+auto operator<<(llvm::raw_ostream& os, HashedValue<std::pair<T, U>> hv)
+    -> llvm::raw_ostream& {
+  os << "hash " << hv.hash << " for pair of ";
+  PrintFullWidthHex(os, hv.v.first);
+  os << " and ";
+  PrintFullWidthHex(os, hv.v.second);
+  return os;
+}
+
+struct Collisions {
+  int total;
+  int median_per_hash;
+  int max_per_hash;
+};
+
+template <int BitBegin, int BitEnd, typename T>
+auto FindBitRangeCollisions(llvm::ArrayRef<HashedValue<T>> hashes)
+    -> Collisions {
+  static_assert(BitBegin < BitEnd);
+  constexpr int BitCount = BitEnd - BitBegin;
+  static_assert(BitCount <= 32);
+  constexpr int BitShift = BitBegin;
+  constexpr uint64_t BitMask = ((1ULL << BitCount) - 1) << BitShift;
+
+  llvm::SmallVector<int> collision_counts;
+  collision_counts.push_back(0);
+  llvm::SmallVector<int> collision_map;
+  collision_map.resize(hashes.size());
+
+  llvm::SmallVector<std::pair<uint32_t, int>> bits_and_indices;
+  bits_and_indices.reserve(hashes.size());
+  for (const auto& [hash, v] : hashes) {
+    CARBON_DCHECK(v == hashes[bits_and_indices.size()].v);
+    auto hash_bits = (static_cast<uint64_t>(hash) & BitMask) >> BitShift;
+    bits_and_indices.push_back(
+        {static_cast<uint32_t>(hash_bits), bits_and_indices.size()});
+  }
+  std::sort(
+      bits_and_indices.begin(), bits_and_indices.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.first < rhs.first; });
+
+  uint32_t prev_hash_bits = bits_and_indices[0].first;
+  int prev_index = bits_and_indices[0].second;
+  bool in_collision = false;
+  int total = 0;
+  int distinct_collisions = 0;
+  for (const auto& [hash_bits, hash_index] :
+       llvm::ArrayRef(bits_and_indices).slice(1)) {
+    if (hash_bits != prev_hash_bits) {
+      prev_hash_bits = hash_bits;
+      prev_index = hash_index;
+      in_collision = false;
+      continue;
+    }
+
+    ++total;
+    if (in_collision) {
+      ++collision_counts.back();
+      collision_map[hash_index] = collision_counts.size() - 1;
+      continue;
+    }
+    in_collision = true;
+    collision_map[hash_index] = collision_counts.size();
+    collision_counts.push_back(1);
+
+    ++distinct_collisions;
+    if (0 && distinct_collisions < 10) {
```

Comment:
Maybe include a comment about how this is expected to be used?
```suggestion
    if (false && distinct_collisions < 10) {
```

Reply:
Debugging code that I've removed. I couldn't find a way to make the output brief enough to be useful to leave in.